qutip/core/dimensions.py

```diff
@@ -343,3 +324,561 @@ def dims_idxs_to_tensor_idxs(dims, indices):
+
+    def __eq__(self, other):
+        return self is other or (
+            type(other) is type(self)
+            and len(self.spaces) == len(other.spaces)
+            and all(self_space == other_space
+                    for self_space, other_space
+                    in zip(self.spaces, other.spaces))
+        )
+
+    def __hash__(self):
+        return hash(self.spaces)
+
+    def __repr__(self):
+        parts_rep = ", ".join(repr(space) for space in self.spaces)
+        return f"Compound({parts_rep})"
+
+    def as_list(self):
+        return sum([space.as_list() for space in self.spaces], [])
+
+    def dims2idx(self, dims):
+        print('Compound', dims)
+        pos = 0
+        step = 1
+        for space, dim in zip(self.spaces[::-1], dims[::-1]):
+            pos += space.dims2idx(dim) * step
+            step *= space.size
+        return pos
+
+    def idx2dims(self, idx):
+        dims = []
+        for space in self.spaces[::-1]:
+            idx, dim = divmod(idx, space.size)
+            dims = space.idx2dims(dim) + dims
+        return dims
+
+    def step(self):
+        steps = []
+        step = 1
+        for space in self.spaces[::-1]:
+            steps = [step * N for N in space.step()] + steps
+            step *= space.size
+        return steps
+
+    def flat(self):
+        return sum([space.flat() for space in self.spaces], [])
+
+    def remove(self, idx):
+        new_spaces = []
+        for space in self.spaces:
+            n_indices = len(space.flat())
+            idx_space = [i for i in idx if i < n_indices]
+            idx = [i-n_indices for i in idx if i >= n_indices]
+            new_space = space.remove(idx_space)
+        if new_spaces:
+            return Compound(*new_spaces)
+        return Field()
+
+    def replace(self, idx, new):
+        new_spaces = []
+        for space in self.spaces:
+            n_indices = len(space.flat())
+            if 0 <= idx < n_indices:
+                new_spaces.append(space.replace(idx, new))
+            else:
+                new_spaces.append(space)
+            idx -= n_indices
+        return Compound(*new_spaces)
+
+
+class SuperSpace(Space):
+    _stored_dims = {}
+
+    def __init__(self, oper, rep='super'):
+        self.oper = oper
+        self.superrep = rep
+        self.size = oper.shape[0] * oper.shape[1]
+        self.issuper = True
+        self._pure_dims = oper._pure_dims
+        self.__setitem__ = _frozen
+
+    def __eq__(self, other):
+        return (
+            self is other
+            or self.oper == other
+            or (
+                type(other) is type(self)
+                and self.oper == other.oper
+                and self.superrep == other.superrep
+            )
+        )
+
+    def __hash__(self):
+        return hash((self.oper, self.superrep))
+
+    def __repr__(self):
+        return f"Super({repr(self.oper)}, rep={self.superrep})"
+
+    def as_list(self):
+        return self.oper.as_list()
```

Comment:
I'm still confused. Why does `SuperSpace(oper)` have the same dims as `oper`? Maybe I am missing something.

Reply:
Not too sure what there is to be confused about, it's what we always had.
`SuperSpace`  is the `oper_dims` as a 1D Space bundled with the `superrep`. 
```
oper_dims = [[2], [2]] = Dimensions(Space, Space)
super_dims = [oper_dims, oper_dims] = Dimensions(SuperSpace(oper_dims, superrep=...), SuperSpace(...))
```
Thus the `SuperSpace` list representation is the same as the oper list representation.

If we say the same as in the sense of `==`, since up to now when we compare dims, we compare the list format, `SuperSpace(oper) == oper` and the check for the `superrep` is dropped.

This is how it works in v4:
```
>>> sup = qt.to_super(qt.destroy(4))
>>> dm = qt.rand_dm(4)
>>> sup(dm) == qt.vector_to_operator(sup * qt.operator_to_vector(dm))
True
# Super being the default representation, this is OK.
>>> sup(dm)
# But this also work, giving different results.
>>> chi = qt.to_chi(qt.destroy(4))
>>> chi(dm)
# So does this
>>> choi = qt.to_choi(qt.destroy(3))
>>> choi(dm)
```

The `superoper.__call__(oper)` check that `superoper.dims[1] == oper.dims`, but since oper does not have a `superrep`, it skip that check.
The `SuperSpace` reproduce that well.

It does look stange and there is certainly some work needed here and with super operators representation in general. Maybe every dm should have a superrep. Or applying super operators on operators only works when `superrep=super`. Or proper conversion of the density matrix should be done, etc. But that will come in another PR.

Reply:
```
oper_dims = [[2], [2]] = Dimensions(Space, Space)
super_dims = [oper_dims, oper_dims] = Dimensions(SuperSpace(oper_dims, superrep=..
```
Oh, I see. `SuperSpace(...)` is only one part of the super operator dimension. I think I missed that. Let's resolve this and just explain it nicely in the documentation when we get there.